Allow to define part of the Hibernate ORM config at runtime

gsmet opened the pull request:
Fixes #12886

gsmet committed:
Allow to define part of the Hibernate ORM config at runtime

gsmet committed:
Fix datasource configuration property

gsmet committed:
Add toString() for BuildTimeSettings and RuntimeSettings

gsmet commented:
@Sanne I'm still making progress on fixing small CI issues (hopefully I'm done with them now, will check the last build as soon as it's finished) but the general idea is there and it relies on the work I did for the Hibernate Search integration.

Also we need to decide which properties we want there. Log for sure. Database generation looks handy too but @emmanuelbernard seems to be against it.

Sanne commented:
Awesome, thanks for exploring this @gsmet . Do you have any data on the impact this might have on native image efficiency?

Sanne commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/StaticInitHibernatePersistenceProviderResolver.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * During the static init phase, we don't access the PersistenceProviderResolver.
+ */
+final class StaticInitHibernatePersistenceProviderResolver implements PersistenceProviderResolver {
```
I'd love to understand the need for this better. Could you elaborate, and maybe evolve the comments for future reference?

gsmet replied:
This is something we shouldn't access during the static init phase as it's only used when creating the entity manager factory.

Thus, to be extra sure we don't access it, I decided to make it extra safe by throwing exceptions.

We now define the `PersistenceProviderResolver` twice: once at the beginning of the static init phase with this one, which shouldn't be used and the second time at the beginning of the runtime phase where we also push the runtime properties.

BTW, it was more or less done this way before i.e. I haven't changed anything really, just made it clearer. Mostly because I now need to push the runtime properties in the constructor. I could have define a setter and rely on the order of operations but I prefer having things immutable and fully constructed.

gsmet commented:
It shouldn't change anything. I haven't changed any infrastructure at all, just reused what we already had.

Typically, we still cannot make runtime properties the config properties used when collecting the metadata. But the ones used after this first phase can be made runtime if we want.

Sanne commented:
> It shouldn't change anything. I haven't changed any infrastructure at all, just reused what we already had.

Before giving the user the option, all code relating to the "export schema" path was dead code (provided it was disabled) - that should have an impact as DCE is quite strong on such aspects.

The mere fact of exposing options is a change.. I suppose this single patch won't be very significant, but making such changes extansively will add-up and I believe it would be best to keep an eye on impact.

gsmet commented:
> Before giving the user the option, all code relating to the "export schema" path was dead code (provided it was disabled) - that should have an impact as DCE is quite strong on such aspects.

I would be surprised if it was really the case given we put a boolean to define something that will be ultimately be resolved again at runtime when reinitializing the services.

I can check the size of the native image though.

gsmet commented:
Size of the native image of `integration-tests  jpa-postgresql`

== Before patch

=== drop-and-create

```
-rwxrwxr-x. 1 gsmet gsmet 75M Oct 26 16:02 target/quarkus-integration-test-jpa-postgresql-999-SNAPSHOT-runner
```

=== none

```
-rwxrwxr-x. 1 gsmet gsmet 75M Oct 26 16:06 target/quarkus-integration-test-jpa-postgresql-999-SNAPSHOT-runner
```

== After patch

=== none (but overridable at runtime)

```
-rwxrwxr-x. 1 gsmet gsmet 75M Oct 26 16:11 target/quarkus-integration-test-jpa-postgresql-999-SNAPSHOT-runner
```

gsmet commented:
I was surprised by the size of the image and I compared to 1.7.5.Final and things do not look good:

```
-rwxrwxr-x. 1 gsmet gsmet 60M Oct 26 16:17 target/quarkus-integration-test-jpa-postgresql-1.7.5.Final-runner
```
I'm pretty sure I ran my earlier runs carefully with the right branch. I'll check what's going on later today.

Sanne commented:
great! thanks for checking.

I'm pretty sure that "giving options" has a detrimental effect in some cases - but it's always hard to predict w/o checking: both the size impact and the fact that the compiler might see / ignore things.

gsmet commented:
1.8.3.Final is fine and the native image is 60 MB but starting with 1.9.0.Final the native image is 75 MB so we have a 15 MB regression in 1.9.0.Final.

Next stop, trying to understand what's going on :).

gsmet commented:
@Sanne so I confirm the penalty has nothing to do with this PR and this PR doesn't have any native executable cost.

I pinged you in the issue and PR I opened to fix the penalty.

gsmet commented:
I still have some native executable issues with Hibernate Reactive. I need to dig more to understand what's going on.

gsmet committed:
Provide better DCE information when Agroal is not around

When Agroal is not around, we should never delegate to the JDBC-based
PersistenceProvider as things will fail, given we don't have a JDBC
datasource around.

Obviously, the test we introduce is not sufficient as having Agroal
around is not enough to have a proper Hibernate ORM setup but that's the
best we can do and should be sufficient in most cases.

gsmet commented:
@Sanne CI passed in my fork so this is ready for review.

The main change since your last check is this commit: https://github.com/quarkusio/quarkus/pull/12949/commits/ca4c90dc54d163567b17592ba0dd5b2330c070b6 .

It properly clean up things for GraalVM if Agroal is not in the classpath.

Sanne commented:
very nice, +1 for the substitution to move forward. FYI Davide's next task is to remove the fallback from Reactive, but I'm glad we don't have to wait for that.

Sanne commented:
I'm trying to build this locally but having a ton of compilation errors from module `quarkus-container-image-openshift-deployment` . Would you have any clue of what could be wrong locally? Clearly it works on CI and I guess for you...

Would prefer to import this in my IDE to review.

gsmet commented:
I have no idea. I gave up on IDE import but the build compiles fine locally with Maven.

Sanne commented:
nevermind, Maven downloaded corrupted jars and cached them.. this will take a while :)

Sanne commented on extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/recording/RecordedState.java:
```diff
@@ -79,4 +82,8 @@ public MultiTenancyStrategy getMultiTenancyStrategy() {
         return isReactive;
     }
+
+    public boolean isFromPersistenceXml() {
```
great idea to add this, I'm going to need it as well.

Sanne commented on extensions/hibernate-reactive/runtime/src/main/java/io/quarkus/hibernate/reactive/runtime/FastBootHibernateReactivePersistenceProviderResolver.java:
```diff
@@ -6,19 +6,24 @@
-            .singletonList(new FastBootHibernateReactivePersistenceProvider());
+    private final List<PersistenceProvider> persistenceProviders;
+
+    public FastBootHibernateReactivePersistenceProviderResolver(HibernateOrmRuntimeConfig hibernateOrmRuntimeConfig) {
```
Very clever! I expected you'd have to bypass the JPA bootstrap, but this seems more effective.

Sanne reviewed (approved):
awesome
